drafts/when-elixirs-performance-becomes-rust-y.md

```diff
@@ -0,0 +1,434 @@
+
+## Introducing Rust
+
+This is where Rust comes in. Rust is a low level language, like C, developed by Mozilla. It's designed to provide better memory safety than C does, while also maintaining solid performance.
+
+There's a library which allows creating Elixir/Erlang NIF's which make use of Rust, called [rustler](https://github.com/rusterlium/rustler). One of its main features is that its safety should prevent the Beam from crashing if something unexpected happens. In the meanwhile, we get to leverage Rust's performance in our Elixir apps when we need it!
+
+### Writing the Rust code
+
+Disclaimer: I don't have much prior knowledge with Rust, I'm sure this code can be improved. We have to start somewhere!
+
+We'll go through the code piece by piece, the total solution will be posted at the end of the article.
+
+The first thing we need to do is add rustler to the project, so let's do that!
+Let's add `{:rustler, "~> 0.21.0"}`, to our dependencies in `mix.exs` and install it with `mix deps.get`.
+
+That'll allow us to run `mix rustler.new`. Let's call our module `PrimeNumbers` (which will call the rust crate `primenumbers` by default, which is fine).
+
+We'll also need to add rustler as compiler in the project, and define our Rust crate.
+
+```Elixir
+def project
+  [
+    compilers: [:rustler, :phoenix, :gettext] ++ Mix.compilers(),
+    rustler_crates: [
+      primenumbers: []
+    ],
+    ... # other project options
+  ]
+```
+
+Let's start writing the Rust code! there will be a rust file at `native/primenumbers/src/lib.rs`, which looks like this currently:
+
+```Rust
+use rustler::{Encoder, Env, Error, Term};
+
+mod atoms {
+    rustler_atoms! {
+        atom ok;
+        atom error;
+        //atom __true__ = "true";
+        //atom __false__ = "false";
+    }
+}
+
+rustler::rustler_export_nifs! {
+    "Elixir.PrimeNumbers",
+    [
+        ("add", 2, add)
+    ],
+    None
+}
+
+fn add<'a>(env: Env<'a>, args: &[Term<'a>]) -> Result<Term<'a>, Error> {
+    let num1: i64 = args[0].decode()?;
+    let num2: i64 = args[1].decode()?;
+
+    Ok((atoms::ok(), num1 + num2).encode(env))
+}
+```
+
+The first statement imports the Encoder, Env, Error and Term types from rustler, so we can make use of them.
+The atoms section declares atoms which will map to Elixir's :ok annd :error atoms, so we can provide an interface similar to elixir.
+
+methods defined in the `rustler::rustler_export_nifs!` statement are made available to Elixir, in the PrimeNumbers module. In the example, add is exported and defined with an arity of 2.
+
+The add function decodes the arguments that are provided, adds them together and yields them back to Elixir as `{:ok, <result>}`. (Although I won't dive deeper into it, rustler makes this possible by implementing the [External Term Format](http://erlang.org/doc/apps/erts/erl_ext_dist.html)).
```

Comment:
I like these small asides that allow readers to dive deeper into the subject, or skip it if you're not interested. Great!